File: Python/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int):
        def rec(left, right, string):
            if len(string) == n * 2:
                arr.append(string)
                return 

            if left < n:
                rec(left + 1, right, string + '(')

            if right < left:
                rec(left, right + 1, string + ')')

        arr = []
        rec(0, 0, '')
        return arr

File: Javascript/22. Generate Parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    function rec(left, right, string) {
        if (string.length === n * 2) {
            arr.push(string);
        }

        if (left < n) {
            rec(left + 1, right, string + "(");
        }
        if (right < left) {
            rec(left, right + 1, string + ")");
        }
    }
    let arr = [];
    rec(0, 0, "");
    return arr;
};